contracts/Marketplace.sol

```diff
@@ -1,74 +1,272 @@
-            return;
-        }
-        serviceContracts[sid] = new Service(msg.sender, sid);
-        services.push(sid);
-        totalServices = totalServices + 1;
-        emit ServiceCreated(sid, serviceContracts[sid]);
-        return serviceContracts[sid];
+    // The following doesn't work but seems cleaner than the previous
+    // Version[] storage versions = new Version[](0);
+    // Offer[] storage offers = new Offer[](0);
+    // Payment[] storage payments = new Payment[](0);
+    // Service storage service = Service({
+    //   sid: sid,
+    //   owner: msg.sender,
+    //   versions: versions,
+    //   offers: offers,
+    //   payments: payments
+    // });
+    // services.push(service);
+
+    emit ServiceCreated(service.owner, service.sid, services.length - 1);
+
+    return services.length - 1;
+  }
+
+  event ServiceCreated(address indexed owner, string indexed sid, uint indexed serviceIndex);
+
+  function changeServiceOwner (string memory sid, address newOwner) public whenNotPaused {
+    uint serviceIndex = getServiceIndex(sid);
+    Service storage service = services[serviceIndex];
+    checkServiceOwner(service);
+    service.owner = newOwner;
+  }
+
+  // Manage Version
+
+  function createServiceVersion (string memory sid, string memory hash, string memory url) public whenNotPaused returns (uint versionIndex) {
+    uint serviceIndex = getServiceIndex(sid);
+    Service storage service = services[serviceIndex];
+    checkServiceOwner(service);
+    service.versions.push(Version({
+      hash: hash,
+      url: url
+    }));
+    return service.versions.length - 1;
+  }
+
+  // Manage Offer
+
+  function createServiceOffer (string memory sid, uint price, address payable seller, bool active) public whenNotPaused returns (uint offerIndex) {
+    uint serviceIndex = getServiceIndex(sid);
+    Service storage service = services[serviceIndex];
+    checkServiceOwner(service);
+    service.offers.push(Offer({
+      price: price,
+      seller: seller,
+      active: active
+    }));
+    return service.offers.length - 1;
+  }
+
+  function editServiceOffer (string memory sid, uint offerIndex, bool active) public whenNotPaused {
+    uint serviceIndex = getServiceIndex(sid);
+    Service storage service = services[serviceIndex];
+    checkServiceOwner(service);
+    service.offers[offerIndex].active = active;
+  }
+
+  // ------------------------------------------------------
+  // Payment
+  // ------------------------------------------------------
+
+  function verify(string memory sid) public view returns (bool isPaid) {
+    uint serviceIndex = getServiceIndex(sid);
+    Service memory service = services[serviceIndex];
+    for (uint i = 0; i < service.payments.length; i++) {
+      if (service.payments[i].purchaser == msg.sender) {
+        return true;
+      }
     }
+    return false;
+  }
+
+  function pay(string memory sid, uint offerIndex) public payable whenNotPaused returns (uint paymentIndex) {
```

Comment:
I don't like the approach where every function search for service based on sid, it will be very expensive....

We can get services count from smart contract as well as each service data for free, then pass just the service index to all those funciton.

Reply:
Updated